server/src/middleware/customer-access-control.ts

```diff
@@ -0,0 +1,67 @@
+/**
+ * Copyright 2020 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {Request, Response, NextFunction} from 'express';
+
+import {
+  UnauthorizedError,
+  ForbiddenError,
+  InternalServerError,
+} from '../utils/http-errors';
+
+/**
+ * Authorizes a customer.
+ * Requires req.authorizedCustomerGpayId to be set in order for this
+ * access control middleware to work.
+ * @params getGpayIdOfAuthorizedCustomer Getter function to get gpayId of
+ * an authorized customer
+ */
+const customerAccessControl = (
+  req: Request,
+  res: Response,
+  next: NextFunction
+) => {
+  try {
+    const decodedIdTokenOfCurrCustomer = req.decodedCustomer;
+    if (
+      decodedIdTokenOfCurrCustomer === undefined ||
+      decodedIdTokenOfCurrCustomer === null
+    ) {
+      throw new UnauthorizedError('Invalid Authorization token format.');
+    }
+
+    const {sub: gpayIdOfCurrCustomer} = decodedIdTokenOfCurrCustomer;
+    if (gpayIdOfCurrCustomer === undefined) {
+      throw new UnauthorizedError('User cannot be identified.');
+    }
+
+    const gpayIdOfAuthorizedCustomer = req.authorizedCustomerGPayId;
+    if (gpayIdOfAuthorizedCustomer === undefined) {
+      throw new InternalServerError('req.authorizedCustomerGPayId is not set.');
+    }
+
+    if (gpayIdOfCurrCustomer !== gpayIdOfAuthorizedCustomer) {
```

Comment:
so gpayIdOfCurrCustomer is from spot identity api. where is gpayIdOfAuthorizedCustomer from?

Reply:
@ifhuang 
`gpayIdOfCurrCustomer` is derived from the **decoded Authorization token** that is placed in Header of requests that requires customer auth
`gpayIdOfAuthorizedCustomer` is derived from the request. So for example you want to edit customer 123's data in the request. `gpayIdOfAuthorizedCustomer` will be the gpayId of customer 123.
We will then check if gpayId of customer 123 matches with the gpay id in the authorization token provided by the customer.
Basically this prevents 1 authorized customer from accessing/modifying another authorized customer's data